[BOLT][AArch64] Allow binary-analysis and heatmap tool to run with pac-ret binaries

bgergely0 opened the pull request:
Quoting `bolt/docs/BinaryAnalysis.md`:
> BOLT cannot currently handle functions with cfi_negate_ra_state correctly, i.e. any binaries built with -mbranch-protection=pac-ret. The scanner is meant to be used on specifically such binaries, so this is a major limitation! Work is going on in PR https://github.com/llvm/llvm-project/pull/120064 to fix this.

For pac-ret analysis to work, we don't actually need the whole work in #120064 to be merged. 

## Some context
The BOLT codebase holds different tools: BOLT itself is used to optimize the layout of a binary, while other tools only need to generate the same internal representation to aid in that process.

The `.cfi_negate_ra_state` Call Frame Instruction needs to be read, tracked, and eventually emitted by tools that produce a binary: this CFI signals to the unwinder if it needs to strip the PAC bits from pointers or not.

This means we can ignore these CFIs in `llvm-bolt-binary-analysis` and the `heatmaptool`.

## Solution
In places where previously BOLT would run to an `llvm_unreachable("Unsupported CFI opcode")` for `OpNegateRAState`, we first check from options which tool is running, and only crash with unreachable when needed.

bgergely0 commented:
PTAL @kbeyls, @paschalis-mpeis

atrosinenko commented on bolt/test/AArch64/dw_cfa_gnu_window_save.test:
```diff
@@ -1,8 +1,8 @@
 
 RUN: yaml2obj %p/Inputs/dw_cfa_gnu_window_save.yaml &> %t.exe
-RUN: llvm-bolt %t.exe -o %t.bolt 2>&1 | FileCheck %s
+RUN not: llvm-bolt %t.exe -o %t.bolt 2>&1 | FileCheck %s
```
`not` is not a modifier of `RUN:` directive, there is a `not` helper binary built alongside other LLVM binaries - it "negates the result" of execution of the command line passed as its arguments (success vs. failure, but IIRC there is a `--crash` mode as well). Here, the command line is the result of expansion of `llvm-bolt %t.exe -o %t.bolt` - it is executed with `stdout` and `stderr` being passed through unchanged to `FileCheck`, but the return code of `llvm-bolt` is "negated".
```suggestion
RUN: not llvm-bolt %t.exe -o %t.bolt 2>&1 | FileCheck %s
```

atrosinenko reviewed (commented):
Thank you very much for unblocking gadget scanner - I think it is a good idea to first fix the "read only mode" and then discuss the general case which is much more complex.

I am rather cautious about matching executable name (though I'm not very familiar with BOLT's core). Maybe it is better to choose the mode of operation based on `opts::BinaryAnalysisMode` and other such options, see [these lines](https://github.com/llvm/llvm-project/blob/e428afdfcf56ccadbbcff16e8fe52e51622baed7/bolt/lib/Rewrite/RewriteInstance.cpp#L747) for example. Then, instead of assigning `this->ToolName`, you could compute this property in constructor and assign it to a boolean field.

bgergely0 commented:
@atrosinenko Thanks for the review.

> Maybe it is better to choose the mode of operation based on opts::BinaryAnalysisMode

Good point, probably better to only change the behaviour for the binary-analysis tool, because the current way it allows all tools besides BOLT. Doing `perf2bolt` when you cannot do `bolt` after isn't much help.

paschalis-mpeis commented on bolt/lib/Core/Exceptions.cpp:
```diff
@@ -634,10 +634,10 @@ bool CFIReaderWriter::fillCFIInfoFor(BinaryFunction &Function) const {
+        // llvm-bolt-binary-analysis does not need this CFI, as it is not
+        // producing a binary. For all other tools, we should refuse processing
+        // the binary, until PAC support is added.
+        if (!opts::BinaryAnalysisMode) {
```
Probably better to add `HeatmapMode` here as well?

In theory one could use `llvm-heatmap` tool on any binary, possibly even after applying some other pgo/post-link tools.

paschalis-mpeis reviewed (commented):
Hey Gergely,

Thanks for your work. I think it should be OK to restrict this to both `llvm-bolt` and `perf2bolt`. As said, the latter is not affected, but then `llvm-bolt` won't be able to run regardless.

paschalis-mpeis reviewed (approved):
Looks good, thanks!

Let's allow 1-2 days in case there are additional comments.

atrosinenko reviewed (commented):
@bgergely0 Thank you for the updates! I have no more comments except for one minor nit-picking (feel free to ignore).

I agree with @paschalis-mpeis, let's wait 1-2 days for others to comment.

paschalis-mpeis commented:
Could also rename the title to something like:

`[BOLT][AArch64] Abort on pac-ret binaries in llvm-bolt and perf2bolt`

bgergely0 commented on bolt/lib/Core/Exceptions.cpp:
```diff
@@ -632,8 +635,15 @@ bool CFIReaderWriter::fillCFIInfoFor(BinaryFunction &Function) const {
+        // they are not producing a binary. For other tools, we should refuse
+        // processing the binary, until PAC support is added.
+        if (!(opts::BinaryAnalysisMode || opts::HeatmapMode)) {
+          BC.errs() << "BOLT-ERROR: pointer authentication is not supported "
+                       "yet. Please compile "
+                       "your target binary using '-mbranch-protection=none'.\n";
```
I left it like it was, because `git clang-format HEAD~1` did not modify it, but agree that it looks funky.

kbeyls reviewed (commented):
Thank you, this mostly looks good to me!
I just had one nit-pick comment.

I would also improve the title of this PR/commit message to be more descriptive, as @paschalis-mpeis recommended.

paschalis-mpeis reviewed (changes requested):
Noting that this new behavior is triggered on 39 tests on AArch64, causing all to fail.
It will have to be addressed before proceeding.

```
Total Discovered Tests: 547
  Skipped          :  13 (2.38%)
  Unsupported      :  81 (14.81%)
  Passed           : 413 (75.50%)
  Expectedly Failed:   1 (0.18%)
  Failed           :  39 (7.13%)
```

bgergely0 commented:
Looks like many binaries have `paciasp`/`autiasp` + the `.cfi_negate_ra_state` in `__do_global_dtors_aux`.
These small unittest targets do not run into `unwindCFIState` and similar functions where `unsupported CFI opcode` would be triggered. Problem is, the `__do_global_dtors_aux` is coming from a precompiled object on the system (as I currently understand it), so the user does not control the flags it is compiled with.

bgergely0 committed:
[BOLT][AArch64] Allow binary-analysis and heatmaptool to run with
pac-ret binaries

- OpNegateRAState support is only needed for tools that
    produce binaries.

bgergely0 commented:
With the new commit I do the check "later", e.g. not in `FillCFIInfo`, but only when BOLT would crash anyways. With this, I can make sure no unittests that passed before my change will fail.

paschalis-mpeis reviewed (approved):
Hey Gergely,

Thanks for fixing the tests. Essentially now the error handling of `OpNegateRAState` is explicit. And your upcoming work will add support through a flag:
- #120064

bgergely0 commented:
Thanks for the approval @paschalis-mpeis!
I don't have write permissions the repo, can you merge this in a few days if others have no concerns?

atrosinenko reviewed (commented):
The patch looks good to me (though I'm not very familiar with BOLT's handling of CFI), with a trivial nit-picking. Thank you!
